Maven dependency collection fails: `java.io.FileNotFoundException: /path/to/dependency-3.x.x.pom.part (No such file or directory)`

### Describe the bug

When building a Quarkus application in a CI environment (private), I am frequently encountering build failures similar to below. It is not caused by a specific dependency, the "No such file or directory" differs every time.

The issue appears to stem from [NonBlockingModelResolutionTaskRunner](https://github.com/quarkusio/quarkus/blob/main/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/NonBlockingModelResolutionTaskRunner.java) and related classes. This has not been an issue with Spring or other frameworks.

### Expected behavior

The error `/home/runner/.m2/repository/io/quarkus/arc/arc-processor/3.19.1/arc-processor-3.19.1.pom.part (No such file or directory)` is quite strange, it's unclear to me what the specific failure is.

If it's a transport failure, could there be better handling or retries? Otherwise, try to make the issue more explicit.

### Actual behavior

The build fails with the aforementioned issue, though like I said the _actual_ issue isn't clear to me.

Some ideas:

- Issue downloading the complete file from the remote repository
- Issue writing the file to disk
- A race condition where repository `$c` says "dependency has been downloaded" despite it returning a 404 because the dependency only exists in repository `$a`.

```
[ERROR] [io.quarkus.bootstrap.resolver.maven.NonBlockingModelResolutionTaskRunner] The following errors were encountered while processing Quarkus application dependencies:
[ERROR] [io.quarkus.bootstrap.resolver.maven.NonBlockingModelResolutionTaskRunner] 1)
 io.quarkus.bootstrap.resolver.maven.DeploymentInjectionException: Failed to collect dependencies for io.quarkus:quarkus-vertx-kotlin-deployment:jar:3.19.1
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver.collectDependencies (ApplicationDependencyResolver.java:831)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$ExtensionDependency.collectDeploymentDeps (ApplicationDependencyResolver.java:911)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$AppDep.collectDeploymentDeps (ApplicationDependencyResolver.java:766)
     at io.quarkus.bootstrap.resolver.maven.NonBlockingModelResolutionTaskRunner.lambda$run$0 (NonBlockingModelResolutionTaskRunner.java:34)
     at java.util.concurrent.CompletableFuture$AsyncRun.run (CompletableFuture.java:1804)
     at java.lang.Thread.run (Thread.java:1583)
 Caused by: org.eclipse.aether.collection.DependencyCollectionException: Failed to collect dependencies at io.quarkus:quarkus-vertx-kotlin-deployment:jar:3.19.1 -> io.quarkus:quarkus-arc-deployment:jar:3.19.1 -> io.quarkus.arc:arc-processor:jar:3.19.1
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies (DefaultDependencyCollector.java:288)
     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies (DefaultRepositorySystem.java:284)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver.collectDependencies (ApplicationDependencyResolver.java:828)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$ExtensionDependency.collectDeploymentDeps (ApplicationDependencyResolver.java:911)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$AppDep.collectDeploymentDeps (ApplicationDependencyResolver.java:766)
     at io.quarkus.bootstrap.resolver.maven.NonBlockingModelResolutionTaskRunner.lambda$run$0 (NonBlockingModelResolutionTaskRunner.java:34)
     at java.util.concurrent.CompletableFuture$AsyncRun.run (CompletableFuture.java:1804)
     at java.lang.Thread.run (Thread.java:1583)
 Caused by: org.eclipse.aether.resolution.ArtifactDescriptorException: Failed to read artifact descriptor for io.quarkus.arc:arc-processor:jar:3.19.1
     at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom (DefaultArtifactDescriptorReader.java:255)
     at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor (DefaultArtifactDescriptorReader.java:171)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.resolveCachedArtifactDescriptor (DefaultDependencyCollector.java:538)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.getArtifactDescriptorResult (DefaultDependencyCollector.java:523)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:410)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:506)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:458)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:506)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:458)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies (DefaultDependencyCollector.java:254)
     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies (DefaultRepositorySystem.java:284)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver.collectDependencies (ApplicationDependencyResolver.java:828)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$ExtensionDependency.collectDeploymentDeps (ApplicationDependencyResolver.java:911)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$AppDep.collectDeploymentDeps (ApplicationDependencyResolver.java:766)
     at io.quarkus.bootstrap.resolver.maven.NonBlockingModelResolutionTaskRunner.lambda$run$0 (NonBlockingModelResolutionTaskRunner.java:34)
     at java.util.concurrent.CompletableFuture$AsyncRun.run (CompletableFuture.java:1804)
     at java.lang.Thread.run (Thread.java:1583)
 Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Could not transfer artifact io.quarkus.arc:arc-processor:pom:3.19.1 from/to central-mirror (https://<redacted>:443/artifactory/oss-maven-central/): /home/runner/.m2/repository/io/quarkus/arc/arc-processor/3.19.1/arc-processor-3.19.1.pom.part (No such file or directory)
     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:425)
     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts (DefaultArtifactResolver.java:229)
     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact (DefaultArtifactResolver.java:207)
     at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom (DefaultArtifactDescriptorReader.java:240)
     at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor (DefaultArtifactDescriptorReader.java:171)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.resolveCachedArtifactDescriptor (DefaultDependencyCollector.java:538)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.getArtifactDescriptorResult (DefaultDependencyCollector.java:523)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:410)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:506)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:458)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:506)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:458)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies (DefaultDependencyCollector.java:254)
     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies (DefaultRepositorySystem.java:284)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver.collectDependencies (ApplicationDependencyResolver.java:828)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$ExtensionDependency.collectDeploymentDeps (ApplicationDependencyResolver.java:911)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$AppDep.collectDeploymentDeps (ApplicationDependencyResolver.java:766)
     at io.quarkus.bootstrap.resolver.maven.NonBlockingModelResolutionTaskRunner.lambda$run$0 (NonBlockingModelResolutionTaskRunner.java:34)
     at java.util.concurrent.CompletableFuture$AsyncRun.run (CompletableFuture.java:1804)
     at java.lang.Thread.run (Thread.java:1583)
 Caused by: org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact io.quarkus.arc:arc-processor:pom:3.19.1 from/to central-mirror (https://<redacted>:443/artifactory/oss-maven-central/): /home/runner/.m2/repository/io/quarkus/arc/arc-processor/3.19.1/arc-processor-3.19.1.pom.part (No such file or directory)
     at org.eclipse.aether.connector.basic.ArtifactTransportListener.transferFailed (ArtifactTransportListener.java:52)
     at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run (BasicRepositoryConnector.java:369)
     at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run (RunnableErrorForwarder.java:75)
     at org.eclipse.aether.connector.basic.BasicRepositoryConnector$DirectExecutor.execute (BasicRepositoryConnector.java:628)
     at org.eclipse.aether.connector.basic.BasicRepositoryConnector.get (BasicRepositoryConnector.java:262)
     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.performDownloads (DefaultArtifactResolver.java:514)
     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:402)
     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts (DefaultArtifactResolver.java:229)
     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact (DefaultArtifactResolver.java:207)
     at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom (DefaultArtifactDescriptorReader.java:240)
     at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor (DefaultArtifactDescriptorReader.java:171)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.resolveCachedArtifactDescriptor (DefaultDependencyCollector.java:538)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.getArtifactDescriptorResult (DefaultDependencyCollector.java:523)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:410)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:506)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:458)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:506)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:458)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
     at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies (DefaultDependencyCollector.java:254)
     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies (DefaultRepositorySystem.java:284)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver.collectDependencies (ApplicationDependencyResolver.java:828)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$ExtensionDependency.collectDeploymentDeps (ApplicationDependencyResolver.java:911)
     at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$AppDep.collectDeploymentDeps (ApplicationDependencyResolver.java:766)
     at io.quarkus.bootstrap.resolver.maven.NonBlockingModelResolutionTaskRunner.lambda$run$0 (NonBlockingModelResolutionTaskRunner.java:34)
     at java.util.concurrent.CompletableFuture$AsyncRun.run (CompletableFuture.java:1804)
     at java.lang.Thread.run (Thread.java:1583)
Caused by: java.io.FileNotFoundException: /home/runner/.m2/repository/io/quarkus/arc/arc-processor/3.19.1/arc-processor-3.19.1.pom.part (No such file or directory)
    at java.io.FileInputStream.open0 (Native Method)
    at java.io.FileInputStream.open (FileInputStream.java:213)
    at java.io.FileInputStream.<init> (FileInputStream.java:152)
    at org.eclipse.aether.internal.impl.DefaultFileProcessor.copy (DefaultFileProcessor.java:163)
    at org.eclipse.aether.internal.impl.DefaultFileProcessor.copy (DefaultFileProcessor.java:151)
    at org.eclipse.aether.internal.impl.DefaultFileProcessor.move (DefaultFileProcessor.java:252)
    at org.eclipse.aether.connector.basic.BasicRepositoryConnector$GetTaskRunner.runTask (BasicRepositoryConnector.java:482)
    at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run (BasicRepositoryConnector.java:364)
    at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run (RunnableErrorForwarder.java:75)
    at org.eclipse.aether.connector.basic.BasicRepositoryConnector$DirectExecutor.execute (BasicRepositoryConnector.java:628)
    at org.eclipse.aether.connector.basic.BasicRepositoryConnector.get (BasicRepositoryConnector.java:262)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.performDownloads (DefaultArtifactResolver.java:514)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:402)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts (DefaultArtifactResolver.java:229)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact (DefaultArtifactResolver.java:207)
    at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom (DefaultArtifactDescriptorReader.java:240)
    at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor (DefaultArtifactDescriptorReader.java:171)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.resolveCachedArtifactDescriptor (DefaultDependencyCollector.java:538)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.getArtifactDescriptorResult (DefaultDependencyCollector.java:523)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:410)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:506)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:458)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:506)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:458)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies (DefaultDependencyCollector.java:254)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies (DefaultRepositorySystem.java:284)
    at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver.collectDependencies (ApplicationDependencyResolver.java:828)
    at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$ExtensionDependency.collectDeploymentDeps (ApplicationDependencyResolver.java:911)
    at io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver$AppDep.collectDeploymentDeps (ApplicationDependencyResolver.java:766)
    at io.quarkus.bootstrap.resolver.maven.NonBlockingModelResolutionTaskRunner.lambda$run$0 (NonBlockingModelResolutionTaskRunner.java:34)
    at java.util.concurrent.CompletableFuture$AsyncRun.run (CompletableFuture.java:1804)
    at java.lang.Thread.run (Thread.java:1583)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  49.536 s
[INFO] Finished at: 2025-02-27T15:50:52Z
[INFO] ------------------------------------------------------------------------
```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

21

### Quarkus version or git rev

 3.23.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

1. We have multiple repositories. There are attempts to download each dependency from multiple places, some of which may fail or return a 404.
2. I am happy to share the `mvn -X` debug log privately. I've elected not to upload it, as I couldn't guarantee that it was free of sensitive information.

--- a/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/PackageAppTestBase.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/PackageAppTestBase.java
@@ -1,9 +1,9 @@
 package io.quarkus.deployment.runnerjar;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
-import static org.junit.jupiter.api.Assertions.fail;
 
 import java.io.BufferedWriter;
 import java.io.IOException;
@@ -41,7 +41,7 @@ import io.quarkus.maven.dependency.ResolvedDependency;
 
 public abstract class PackageAppTestBase extends BootstrapTestBase {
 
-    private static final String LIB_PREFIX = "lib/";
+    private static final String LIB_BOOT_PREFIX = "lib/boot/";
     private static final String MAIN_CLS = "io.quarkus.bootstrap.runner.QuarkusEntryPoint";
 
     protected List<String> expectedLib = new ArrayList<>();
@@ -209,44 +209,16 @@ public abstract class PackageAppTestBase extends BootstrapTestBase {
                     .toArray(String[]::new);
             assertEquals(actualBootLib.size(), cpEntries.length);
             for (String entry : cpEntries) {
-                assertTrue(entry.startsWith(LIB_PREFIX));
-                assertTrue(actualBootLib.contains(entry.substring(LIB_PREFIX.length())));
+                assertThat(entry).startsWith(LIB_BOOT_PREFIX);
+                String entryFile = entry.substring(LIB_BOOT_PREFIX.length());
+                assertThat(actualBootLib).contains(entryFile);
             }
         }
+        assertLibDirectoryContent(actualMainLib);
+    }
 
-        List<String> missingEntries = List.of();
-        for (String entry : expectedLib) {
-            if (!actualMainLib.remove(entry)) {
-                if (missingEntries.isEmpty()) {
-                    missingEntries = new ArrayList<>();
-                }
-                missingEntries.add(entry);
-            }
-        }
-
-        StringBuilder buf = null;
-        if (!missingEntries.isEmpty()) {
-            buf = new StringBuilder();
-            buf.append("Missing entries: ").append(missingEntries.get(0));
-            for (int i = 1; i < missingEntries.size(); ++i) {
-                buf.append(", ").append(missingEntries.get(i));
-            }
-        }
-        if (!actualMainLib.isEmpty()) {
-            if (buf == null) {
-                buf = new StringBuilder();
-            } else {
-                buf.append("; ");
-            }
-            final Iterator<String> i = actualMainLib.iterator();
-            buf.append("Extra entries: ").append(i.next());
-            while (i.hasNext()) {
-                buf.append(", ").append(i.next());
-            }
-        }
-        if (buf != null) {
-            fail(buf.toString());
-        }
+    protected void assertLibDirectoryContent(Set<String> actualMainLib) {
+        assertThat(actualMainLib).containsExactlyInAnyOrderElementsOf(expectedLib);
     }
 
     protected Set<String> getDirContent(final Path dir) throws IOException {

--- a/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/RetryResolutionOnMissingTmpFileErrorsTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/RetryResolutionOnMissingTmpFileErrorsTest.java
@@ -1,0 +1,280 @@
+package io.quarkus.deployment.runnerjar;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.BufferedWriter;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.maven.settings.Profile;
+import org.apache.maven.settings.Repository;
+import org.apache.maven.settings.RepositoryPolicy;
+import org.apache.maven.settings.Settings;
+import org.apache.maven.settings.io.DefaultSettingsWriter;
+import org.eclipse.aether.AbstractRepositoryListener;
+import org.eclipse.aether.DefaultRepositorySystemSession;
+import org.eclipse.aether.RepositoryEvent;
+import org.eclipse.aether.RequestTrace;
+import org.eclipse.aether.collection.CollectRequest;
+
+import io.quarkus.bootstrap.model.ApplicationModel;
+import io.quarkus.bootstrap.resolver.TsArtifact;
+import io.quarkus.bootstrap.resolver.maven.BootstrapMavenContext;
+import io.quarkus.bootstrap.resolver.maven.BootstrapMavenException;
+import io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver;
+import io.quarkus.bootstrap.resolver.maven.workspace.LocalProject;
+import io.quarkus.bootstrap.util.DependencyUtils;
+import io.quarkus.maven.dependency.ArtifactCoords;
+import io.quarkus.maven.dependency.DependencyFlags;
+
+public class RetryResolutionOnMissingTmpFileErrorsTest extends BootstrapFromOriginalJarTestBase {
+
+    private static final String QUARKUS_VERSION = System.getProperty("project.version");
+
+    private static class TempFileCleaner implements Runnable {
+
+        private final Path targetRepositoryFile;
+
+        private TempFileCleaner(Path targetRepositoryFile) {
+            this.targetRepositoryFile = targetRepositoryFile;
+        }
+
+        @Override
+        public void run() {
+            final String targetFileName = targetRepositoryFile.getFileName().toString();
+            final Path artifactDir = targetRepositoryFile.getParent();
+            boolean terminate = false;
+            while (!terminate) {
+                if (Files.exists(artifactDir)) {
+                    try (DirectoryStream<Path> stream = Files.newDirectoryStream(artifactDir)) {
+                        for (var p : stream) {
+                            final String fileName = p.getFileName().toString();
+                            if (fileName.equals(targetFileName)) {
+                                terminate = true;
+                            } else if (fileName.endsWith(".tmp") || fileName.endsWith(".part")) {
+                                // delete in-progress files to fail artifact resolution
+                                if (Files.deleteIfExists(p)) {
+                                    System.out.println("Deleted " + p.getFileName());
+                                    terminate = true;
+                                }
+                            }
+                        }
+                    } catch (Exception e) {
+                        e.printStackTrace();
+                    }
+                }
+                //                try {
+                //                    Thread.sleep(50);
+                //                } catch (InterruptedException e) {
+                //                }
+                if (!terminate) {
+                    terminate = Files.exists(targetRepositoryFile);
+                }
+            }
+        }
+    }
+
+    final AtomicBoolean sawFileMissingErrors = new AtomicBoolean(false);
+
+    @Override
+    protected MavenArtifactResolver newArtifactResolver(LocalProject currentProject) throws BootstrapMavenException {
+
+        /**
+         * This is the tricky part.
+         * Here we initialize a Maven artifact resolver to be used by Quarkus to resolve the ApplicationModel.
+         *
+         * First, it will capture events indicating that certain artifacts will be resolved. For non-SNAPSHOT
+         * artifacts, we register tasks that will be scanning those artifact directories in the local Maven repository
+         * trying to capture temporary (in-progress) files that the "Maven resolver" writes artifact content to
+         * before renaming them to the final artifact files.
+         * Once those in-progress files are detected, they are removed. This will cause Quarkus deployment dependency
+         * CollectRequest to fail, which the {@link io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver}
+         * is expected to re-try.
+         *
+         * This is done exclusively for deployment dependencies, since this is what is parallelized
+         * in the {@link io.quarkus.bootstrap.resolver.maven.ApplicationDependencyResolver}.
+         *
+         * The test uses a custom location for a local Maven repository. To be able to resolve locally built Quarkus artifacts,
+         * we create custom settings an active profile where the current local repository is added as remote repository.
+         */
+
+        var originalMvnCtx = new BootstrapMavenContext((BootstrapMavenContext.config().setWorkspaceDiscovery(false)));
+        final Settings settings = originalMvnCtx.getEffectiveSettings();
+        final String originalLocalRepo = originalMvnCtx.getLocalRepo();
+
+        Path settingsXml;
+        try {
+            Profile profile = new Profile();
+            settings.addActiveProfile("original-local");
+            profile.setId("original-local");
+
+            final Repository repo = new Repository();
+            repo.setId("original-local");
+            repo.setLayout("default");
+            repo.setUrl(Path.of(originalLocalRepo).toUri().toURL().toExternalForm());
+            RepositoryPolicy releases = new RepositoryPolicy();
+            releases.setEnabled(false);
+            releases.setChecksumPolicy("ignore");
+            releases.setUpdatePolicy("never");
+            repo.setReleases(releases);
+            RepositoryPolicy snapshots = new RepositoryPolicy();
+            snapshots.setEnabled(true);
+            snapshots.setChecksumPolicy("ignore");
+            snapshots.setUpdatePolicy("never");
+            repo.setSnapshots(snapshots);
+
+            profile.addRepository(repo);
+            profile.addPluginRepository(repo);
+
+            settings.addProfile(profile);
+
+            settingsXml = workDir.resolve("settings.xml");
+            try (BufferedWriter writer = Files.newBufferedWriter(settingsXml)) {
+                new DefaultSettingsWriter().write(writer, Map.of(), settings);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to create test settings.xml", e);
+        }
+
+        var ctx = new BootstrapMavenContext(BootstrapMavenContext.config()
+                // A bit of a peculiar detail, there appears to be a difference between running`mvn test`
+                // and `mvn test -Dtest=RetryResolutionOnTmpFilesMissingTest`.
+                // Unless the current project is set to the module directory, the test belongs to, `mvn test` will fail
+                // while `mvn test -Dtest=RetryResolutionOnTmpFilesMissingTest` will not.
+                .setCurrentProject(Path.of("").normalize().toAbsolutePath().toString())
+                .setWorkspaceDiscovery(true)
+                .setArtifactTransferLogging(false)
+                .setUserSettings(settingsXml.toFile())
+                .setLocalRepository(getLocalRepoHome().toString()));
+
+        var session = new DefaultRepositorySystemSession(ctx.getRepositorySystemSession());
+        final Map<ArtifactCoords, CompletableFuture<?>> tmpFileCleaners = new ConcurrentHashMap<>();
+        final Set<String> ignoredGroupIds = Set.of("org.acme", "io.quarkus.bootstrap.test");
+        session.setRepositoryListener(new AbstractRepositoryListener() {
+            @Override
+            public void artifactResolving(RepositoryEvent event) {
+                // we are looking exclusively for deployment dependency requests
+                if (!isPartOfCollectingDeploymentDeps(event)) {
+                    return;
+                }
+                var a = event.getArtifact();
+                // delete non-snapshot artifacts to trigger re-download but only once
+                if (!ignoredGroupIds.contains(a.getGroupId()) && !a.isSnapshot()) {
+                    tmpFileCleaners.computeIfAbsent(
+                            DependencyUtils.getCoords(a),
+                            k -> {
+                                final Path targetRepositoryFile = event.getSession().getLocalRepository().getBasedir().toPath()
+                                        .resolve(event.getSession().getLocalRepositoryManager().getPathForLocalArtifact(a));
+                                try {
+                                    if (Files.deleteIfExists(targetRepositoryFile)) {
+                                        System.out.println("To be (re)resolved " + event.getArtifact());
+                                    }
+                                } catch (IOException e) {
+                                }
+                                return CompletableFuture.runAsync(new TempFileCleaner(targetRepositoryFile));
+                            });
+                }
+            }
+
+            @Override
+            public void artifactResolved(RepositoryEvent event) {
+                // just to catch the fact that some requests failed due to missing files
+                if (!sawFileMissingErrors.get() && !event.getExceptions().isEmpty()) {
+                    for (var e : event.getExceptions()) {
+                        if (isCausedByMissingFile(e)) {
+                            sawFileMissingErrors.set(true);
+                            break;
+                        }
+                    }
+                }
+            }
+
+            @Override
+            public void artifactDownloading(RepositoryEvent event) {
+                final ArtifactCoords coords = DependencyUtils.getCoords(event.getArtifact());
+                StringBuilder sb = new StringBuilder();
+                sb.append(tmpFileCleaners.containsKey(coords) ? "Re-downloading" : "Downloading")
+                        .append(" from ").append(event.getRepository().getId()).append(": ").append(coords.toCompactCoords());
+                System.out.println(sb);
+            }
+        });
+
+        return MavenArtifactResolver.builder()
+                .setRepositorySystem(ctx.getRepositorySystem())
+                .setRepositorySystemSession(session)
+                .setRemoteRepositories(ctx.getRemoteRepositories())
+                .setRemoteRepositoryManager(ctx.getRemoteRepositoryManager())
+                .setSettingsDecrypter(ctx.getSettingsDecrypter())
+                .build();
+    }
+
+    @Override
+    protected TsArtifact composeApplication() {
+        assertThat(QUARKUS_VERSION).isNotNull();
+        return TsArtifact.jar("app")
+                .addManagedDependency(platformDescriptor())
+                .addManagedDependency(platformProperties())
+                .addDependency(TsArtifact.jar("io.quarkus", "quarkus-core", QUARKUS_VERSION));
+    }
+
+    @Override
+    protected void assertAppModel(ApplicationModel model) throws Exception {
+        // make sure there were resolution failures related to missing files
+        assertThat(sawFileMissingErrors).isTrue();
+
+        // basic model assertion
+        Set<ArtifactCoords> topCoords = new HashSet<>(1);
+        for (var topExt : model.getDependencies(DependencyFlags.TOP_LEVEL_RUNTIME_EXTENSION_ARTIFACT)) {
+            topCoords.add(ArtifactCoords.of(topExt.getGroupId(), topExt.getArtifactId(), topExt.getClassifier(),
+                    topExt.getType(), topExt.getVersion()));
+        }
+        assertThat(topCoords).containsExactly(ArtifactCoords.jar("io.quarkus", "quarkus-core", QUARKUS_VERSION));
+    }
+
+    @Override
+    protected void assertLibDirectoryContent(Set<String> actualMainLib) {
+        // skipping, since it involves iterating over quarkus-core dependencies
+    }
+
+    private static boolean isCausedByMissingFile(Exception e) {
+        Throwable t = e;
+        while (t != null) {
+            // It looks like in Maven 3.9 it's NoSuchFileException, while in Maven 3.8 it's FileNotFoundException
+            if (t instanceof NoSuchFileException || t instanceof FileNotFoundException) {
+                return true;
+            }
+            t = t.getCause();
+        }
+        return false;
+    }
+
+    private static boolean isPartOfCollectingDeploymentDeps(RepositoryEvent event) {
+        RequestTrace requestTrace = getRootRequest(event);
+        if (requestTrace.getData() instanceof CollectRequest cr) {
+            if (cr.getRootArtifact().getArtifactId().endsWith("-deployment")) {
+                return true;
+            }
+        }
+        ;
+        return false;
+    }
+
+    private static RequestTrace getRootRequest(RepositoryEvent event) {
+        RequestTrace requestTrace = event.getTrace();
+        RequestTrace parent;
+        while ((parent = requestTrace.getParent()) != null) {
+            requestTrace = parent;
+        }
+        return requestTrace;
+    }
+}

--- a/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/ResolverSetupCleanup.java
+++ b/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/ResolverSetupCleanup.java
@@ -138,6 +138,10 @@ public class ResolverSetupCleanup {
         return settingsXml;
     }
 
+    protected Path getLocalRepoHome() {
+        return localRepoHome;
+    }
+
     protected boolean cleanWorkDir() {
         return true;
     }

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/ApplicationDependencyResolver.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/ApplicationDependencyResolver.java
@@ -7,17 +7,21 @@ import static io.quarkus.bootstrap.util.DependencyUtils.hasWinner;
 import static io.quarkus.bootstrap.util.DependencyUtils.newDependencyBuilder;
 
 import java.io.BufferedReader;
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedDeque;
 import java.util.function.BiConsumer;
@@ -816,18 +820,71 @@ public class ApplicationDependencyResolver {
 
     private DependencyNode collectDependencies(Artifact artifact, Collection<Exclusion> exclusions,
             List<RemoteRepository> repos) {
-        DependencyNode root;
+        final CollectRequest collectRequest = getCollectRequest(artifact, exclusions, repos);
+        DependencyNode root = null;
         try {
             root = resolver.getSystem()
-                    .collectDependencies(resolver.getSession(), getCollectRequest(artifact, exclusions, repos))
+                    .collectDependencies(resolver.getSession(), collectRequest)
                     .getRoot();
         } catch (DependencyCollectionException e) {
-            throw new DeploymentInjectionException("Failed to collect dependencies for " + artifact, e);
+            // It could happen, especially in Maven 3.8, that multiple threads could end up writing/reading
+            // the same temporary files while resolving the same artifact. Once one of the threads completes
+            // resolving the artifact, the temporary file will be renamed to the target artifact file
+            // and the other thread will fail with one of the file-not-found exceptions.
+            // In this case, we simply re-try the collect request, which should now pick up the already resolved artifact.
+            String missingFile = getMissingFileOrNull(e);
+            if (missingFile != null) {
+                Set<String> missingFiles = new HashSet<>();
+                while (missingFile != null) {
+                    if (missingFiles.add(missingFile)) {
+                        log.debugf("Re-trying the collect request for %s due to missing %s", artifact, missingFile);
+                        try {
+                            root = resolver.getSystem()
+                                    .collectDependencies(resolver.getSession(), collectRequest)
+                                    .getRoot();
+                            break;
+                        } catch (DependencyCollectionException dce) {
+                            missingFile = getMissingFileOrNull(dce);
+                        }
+                    } else {
+                        // if it's the second time it's missing, we give up
+                        throw wrapInDeploymentInjectionException(artifact, e);
+                    }
+                }
+            }
+            if (root == null) {
+                throw wrapInDeploymentInjectionException(artifact, e);
+            }
         }
         if (root.getChildren().size() != 1) {
             throw new DeploymentInjectionException("Only one child expected but got " + root.getChildren());
         }
         return root.getChildren().get(0);
+    }
+
+    private static DeploymentInjectionException wrapInDeploymentInjectionException(Artifact artifact, Exception e) {
+        return new DeploymentInjectionException("Failed to collect dependencies for " + artifact, e);
+    }
+
+    /**
+     * Checks whether the cause of this exception a kind of no-such-file exception and returns the file that was missing.
+     *
+     * @param dce top level exception
+     * @return missing file that was the cause or null, if the cause was different
+     */
+    private static String getMissingFileOrNull(DependencyCollectionException dce) {
+        Throwable t = dce.getCause();
+        while (t != null) {
+            var cause = t.getCause();
+            // It looks like in Maven 3.9 it's NoSuchFileException, while in Maven 3.8 it's FileNotFoundException
+            if (cause instanceof NoSuchFileException e) {
+                return e.getFile();
+            } else if (cause instanceof FileNotFoundException) {
+                return cause.getMessage();
+            }
+            t = cause;
+        }
+        return null;
     }
 
     private CollectRequest getCollectRequest(Artifact artifact, Collection<Exclusion> exclusions,

